subsys/net/lib/wifi_credentials/wifi_credentials.c

```diff
@@ -0,0 +1,444 @@
+#include <zephyr/settings/settings.h>
+
+BUILD_ASSERT(ENTRY_MAX_LEN <= SETTINGS_MAX_VAL_LEN);
+
+/**
+ * @brief WiFi credentials settings handler
+ * @note gets called when settings are loaded
+ *
+ * @param key credential index
+ * @param len length of entry
+ * @param read_cb read callback
+ * @param cb_arg argument for read callback
+ * @return 0 on success, -ERRNO on error
+ */
+static int settings_set(const char *key, size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	/* key validation */
+	if (!key) {
+		return -EINVAL;
+	}
+	int idx = atoi(key);
+
+	if ((idx == 0 && strcmp(key, "0") != 0) ||
+	     idx >= CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES) {
+		return -EINVAL;
+	}
+	if (len < sizeof(struct wifi_credentials_header)) {
+		return -EINVAL;
+	}
+	uint8_t buf[ENTRY_MAX_LEN];
+	size_t length_read = read_cb(cb_arg, buf, ARRAY_SIZE(buf));
+
+	/* value validation */
+	if (length_read < len) {
+		return -ENODATA;
+	}
+
+	load_entry(idx, buf, length_read);
+	return 0;
+}
+
+int wifi_credentials_pickle_entry(size_t idx, const uint8_t *buf, size_t buf_len)
+{
+	char settings_name_buf[16] = {0};
+
+	snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+	return settings_save_one(settings_name_buf, buf, buf_len);
+}
+
+int wifi_credentials_delete_entry(size_t idx)
+{
+	char settings_name_buf[16] = {0};
+
+	snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+	return settings_delete(settings_name_buf);
+}
+
+struct unpickle_helper_arg {
+	uint8_t *buf;
+	size_t buf_len;
+};
+
+static int unpickle_helper(const char *key, size_t len,
+			   settings_read_cb read_cb,
+			   void *cb_arg, void *param)
+{
+	struct unpickle_helper_arg *arg = param;
+
+	if (len < sizeof(struct wifi_credentials_header)) {
+		return -EINVAL;
+	}
+	size_t length_read = read_cb(cb_arg, arg->buf, arg->buf_len);
+
+	/* value validation */
+	if (length_read < len) {
+		return -ENODATA;
+	}
+	return 0;
+}
+
+int wifi_credentials_unpickle_entry(size_t idx, uint8_t *buf, size_t buf_len)
+{
+	struct unpickle_helper_arg arg = {
+		.buf = buf,
+		.buf_len = buf_len
+	};
+	return settings_load_subtree_direct("wifi_cred", unpickle_helper, &arg);
+}
+
+SETTINGS_STATIC_HANDLER_DEFINE(wifi_credential, "wifi_cred", NULL, settings_set, NULL, NULL);
+#endif /* defined(CONFIG_WIFI_CREDENTIALS_BACKEND_SETTINGS) */
+
+#if defined(CONFIG_WIFI_CREDENTIALS_BACKEND_PSA)
+#include "psa/protected_storage.h"
+
+static int init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	psa_status_t ret;
+	uint8_t buf[256];
+
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		size_t length_read = 0;
+
+		ret = psa_ps_get(i + CONFIG_WIFI_CREDENTIALS_BACKEND_PSA_UID_OFFSET,
+				 0, ARRAY_SIZE(buf), buf, &length_read);
+		if (ret == PSA_SUCCESS && length_read >= sizeof(struct wifi_credentials_header)) {
+			load_entry(i, buf, length_read);
+		}
+	}
+	return 0;
+}
+
+SYS_INIT(init, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```

Comment:
I'd move this to the end of PSA backend block. Might be hard to find between functions.

Reply:
done